chain/network/src/routing.rs

```diff
@@ -752,106 +766,172 @@ pub struct ProcessEdgeResult {
 #[derive(Debug)]
 pub struct RoutingTableInfo {
     pub account_peers: HashMap<AccountId, PeerId>,
-    pub peer_forwarding: HashMap<PeerId, HashSet<PeerId>>,
+    pub peer_forwarding: HashMap<PeerId, Vec<PeerId>>,
 }
 
 #[derive(Clone)]
 pub struct Graph {
     pub source: PeerId,
-    adjacency: HashMap<PeerId, HashSet<PeerId>>,
+    source_id: usize,
+    p2id: HashMap<PeerId, usize>,
+    id2p: Vec<PeerId>,
+    used: Vec<bool>,
+    unused: Vec<usize>,
```

Comment:
If we have some kind of utils modue, it might make sense to extract these three fields into a `struct NeArena<T>`, such that they are independent of the graph details. No strong preference either way here.

Reply:
I implemented this structure already we can bring it back: https://github.com/near/nearcore/blob/3749f2b640a7bad48e1843c8c4b99a0a91465b14/chain/network/src/freelist.rs

@bowenwang1996 The question is what would we do with that code.
a) leave it as it is in chain/network/src/freelist.rs
b) move it to crates.io, have someone finish it
c) rewrite Arena to not uses indexes add it to our code
d) rewrite Arena to not uses indexes and publish it
e) do nothing

Reply:
If we have some free time to spend, we can optimize the performance of this code further by another 2-4x. It would probably benefit us more.

Reply:
let's not do that for now. It is not a priority